src/common/buffer.cc

```diff
@@ -844,14 +844,8 @@ using namespace ceph;
     //     << " (p_off " << p_off << " in " << p->length() << ")"
     //     << std::endl;
 
-    p_off += o;
-
-    if (!o) {
-      return;
-    }
-    while (p_off > 0) {
-      if (p == ls->end())
-        throw end_of_buffer();
+    p_off +=o;
+    while (p != ls->end()) {
```

Comment:
@rzarzynski i'd suggest something uglier:
```c++
p_off += o;
off += o;
for (;;) {
  if (p != ls->end()) {
    if (p_off >= p->length()) {
      // skip this buffer
      p_off -= p->length();
      ++p;
    } else {
      // somewhere in this buffer!
      break;
    }
  } else if (p_off) {
    throw end_of_buffer();
  } else {
    // reaching the end
    break;
  }
}
```

in this change:

- use prefix increment operator for non-trivial iterator
- avoid unnecessary check for `end_of_buffer`
- increment `off` at the beginning to improve the readability and to help the compiler to emit more efficient assembly, because x86's `add` 's two operands cannot be both memory locations.

Reply:
Hmm, `unittest_bufferlist` signalizes two failures:
```
$ ./bin/unittest_bufferlist --gtest_filter=BufferListIterator.copy\*
Note: Google Test filter = BufferListIterator.copy*
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from BufferListIterator
[ RUN      ] BufferListIterator.copy
unknown file: Failure
C++ exception with description "buffer::end_of_buffer" thrown in the test body.
[  FAILED  ] BufferListIterator.copy (0 ms)
[ RUN      ] BufferListIterator.copy_in
unknown file: Failure
C++ exception with description "buffer::end_of_buffer" thrown in the test body.
[  FAILED  ] BufferListIterator.copy_in (0 ms)
[----------] 2 tests from BufferListIterator (0 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (0 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] BufferListIterator.copy
[  FAILED  ] BufferListIterator.copy_in

 2 FAILED TESTS
```
for following patch:
```diff
--- a/src/common/buffer.cc
+++ b/src/common/buffer.cc
@@ -734,21 +734,25 @@ static ceph::spinlock debug_lock;
     //     << " (p_off " << p_off << " in " << p->length() << ")"
     //     << std::endl;
 
-    p_off +=o;
-    while (p != ls->end()) {
-      if (p_off >= p->length()) {
-        // skip this buffer
-        p_off -= p->length();
-        p++;
+    p_off += o;
+    off += o;
+    for (;;) {
+      if (p != ls->end()) {
+       if (p_off >= p->length()) {
+         // skip this buffer
+         p_off -= p->length();
+         ++p;
+       } else {
+         // somewhere in this buffer!
+         break;
+       }
+      } else if (p_off) {
+       throw end_of_buffer();
       } else {
-        // somewhere in this buffer!
-        break;
+       // reaching the end
+       break;
       }
     }
-    if (p == ls->end() && p_off) {
-      throw end_of_buffer();
-    }
-    off += o;
   }
 
   template<bool is_const>
```

I propose to optimize `::advance()` later. It seems that the changes don't mess with API nor ABI.